lib/ts/ink_align.h

```diff
@@ -25,67 +25,78 @@
-  uint8_t byte[8];
-  int32_t i32[2];
-  uint32_t u32[2];
-  int64_t i64;
-  uint64_t u64;
-  ink_time_t i_time;
-};
+#include <memory>
+
+//
+// workalike to std::align() ... though it's not a great API
+//
+template <typename T_OBJ>
+inline T_OBJ *ats_align(std::size_t alignment, std::size_t size, T_OBJ *&ptr, std::size_t &space)
+{
+  intptr_t v = reinterpret_cast<intptr_t>(ptr);
+  ptrdiff_t vv = ((v-1) | (alignment-1)) + 1 - v;
+  if ( size + vv > space ) {
+    return nullptr;
+  }
+  space -= vv;
+  return (ptr = reinterpret_cast<T_OBJ*>(v + vv));
+}
 
 /**
  * Alignment macros
  */
 
 #define INK_MIN_ALIGN 8
 /* INK_ALIGN() is only to be used to align on a power of 2 boundary */
-#define INK_ALIGN(size, boundary) (((size) + ((boundary)-1)) & ~((boundary)-1))
+#define INK_ALIGN(size, boundary)  aligned_spacing(size,boundary)
 
 /** Default alignment */
 #define INK_ALIGN_DEFAULT(size) INK_ALIGN(size, INK_MIN_ALIGN)
 
+static inline size_t
+aligned_spacing(size_t len, size_t block=INK_MIN_ALIGN)
+{
+    void *ptr = static_cast<char*>(nullptr) + len; // pointer from zero
+    // next size >= len to get next a new aligned block
+    return static_cast<const char*>( ats_align(block, 0, ptr, block) ) 
+                   - static_cast<const char*>(nullptr);
+}
+
+//
+// Move a pointer backward until it meets an alignment width one before the next.
+//
+static inline void *
+align_pointer_backward(const void *pointer_, size_t alignment)
+{
+    void *bptr = reinterpret_cast<void*>( 0 - reinterpret_cast<intptr_t>(pointer_));
+    // find next aligned ptr if address was "negative" version of original
+    bptr = ats_align(alignment, 0, bptr, alignment); 
+    // re-negate and return
+    return reinterpret_cast<void*>( 0 - reinterpret_cast<intptr_t>(bptr) );
+}
+
+
 //
 // Move a pointer forward until it meets the alignment width.
 //
 static inline void *
 align_pointer_forward(const void *pointer_, size_t alignment)
 {
-  char *pointer = (char *)pointer_;
-  //
-  // Round up alignment..
-  //
-  pointer = (char *)INK_ALIGN((ptrdiff_t)pointer, alignment);
-
-  return (void *)pointer;
+    // next aligned zero length block .. equal or after 
+    return ats_align(alignment, 0, const_cast<void*&>(pointer_), alignment); 
 }
 
 //
 // Move a pointer forward until it meets the alignment width specified,
 // and zero out the contents of the space you're skipping over.
 //
 static inline void *
-align_pointer_forward_and_zero(const void *pointer_, size_t alignment)
+align_pointer_forward_and_zero(void *pointer_, size_t alignment)
 {
-  char *pointer = (char *)pointer_;
-  char *aligned = (char *)INK_ALIGN((ptrdiff_t)pointer, alignment);
-  //
-  // Fill the skippings..
-  //
-  while (pointer < aligned) {
-    *pointer = 0;
-    pointer++;
-  }
-
-  return (void *)aligned;
+    size_t left = alignment;
+    void *aptr = ats_align(alignment, 0, pointer_, left); 
+    memset( pointer_, '\0', alignment - left ); // zero bytes before new block
```

Comment:
I'm pretty sure this doesn't work. `pointer_` is changed by `ats_align` in the return, so it and `aptr` will always be the same. The previous code seems both clearer and correct in this case.